mlabs/src/Mlabs/EfficientNFT/Marketplace.hs

```diff
@@ -2,29 +2,37 @@
 {-# LANGUAGE ImportQualifiedPost #-}
 {-# LANGUAGE TypeApplications #-}
 
-module Mlabs.EfficientNFT.Marketplace (
-  mkPolicy,
-  MintAct (MintToken, ChangePric, ChangeOwner),
-  OwnerData (..),
-  PlatformConfig (..),
-  policy,
-) where
+module Mlabs.EfficientNFT.Marketplace (mkValidator) where
 
-import Data.ByteString (ByteString)
-import Ledger qualified
-import Ledger.Crypto (PubKeyHash)
+import Ledger (
+  ScriptContext,
+  TxInInfo (txInInfoResolved),
+  TxInfo (txInfoOutputs),
+  findOwnInput,
+  scriptContextTxInfo,
+  txOutValue,
+ )
 import Ledger.Value qualified as Value
-import Plutus.V1.Ledger.Scripts qualified as Scripts
-import PlutusTx.Natural (Natural)
+import PlutusTx.Prelude
 
--- TODO
-mkValidator ::
-  TxOutRef ->
-  PubKeyHash ->
-  Natural ->
-  PlatformConfig ->
-  BuiltinData ->
-  BuiltinData ->
-  ScriptContext ->
-  Bool
-mkValidator oref authorPkh royalty platformConfig mintAct ctx = False
+-- | An escrow-like validator, that holds an NFT until sold or pulled out
+{-# INLINEABLE mkValidator #-}
+mkValidator :: BuiltinData -> ScriptContext -> Bool
+mkValidator _ ctx =
+  traceIfFalse "Tokens can only be redeemed when the policy allows a remint" checkRemint
+  where
+    !info = scriptContextTxInfo ctx
+    -- Check if each token from the current input is reminted
+    checkRemint =
+      let !inputVals =
```

Comment:
this should check that there are exactly 2 coins in the escrow (ADA, and the CurrencySymbol we can then use to look in outouts and also ensure that there are no additional coins in the utxo on the output as well. 

Otherwise the escrow is vulnerable to injection of random coins into it which will end up catastrophic if the coin injected can be re-minted.

Reply:
other comment lays out a simpler and more efficient plan

Reply:
just to clarify: we need to check if the output utxo, corresponding to our input has only the token and some ada, right?
I am imagining scenarios, when there are multiple NFTs reminted in one tx (many tokens bought at once), so we must match the inputs to the outputs.
At that point however I don't see how would putting the currency symbol to the datum simplify things.

I was also playing with the idea of a parameterised validator, essentially separate address for each token. But I guess the problem remains the same, and the escrow validator would be tightly coupled to the token, making it less flexible in the future.

Reply:
Or is it enough, if we check the inputs?